climada/util/calibrate/base.py

```diff
@@ -0,0 +1,399 @@
+        here via the ``assign_centroids`` parameter, to avoid assigning them each time
+        the impact is calculated).
+    align_kwds : Mapping (str, Any), optional
+        Keyword arguments to ``pandas.DataFrame.align`` for aligning the :py:attr:`data`
+        with the data frame returned by :py:attr:`impact_to_dataframe`. By default,
+        both axes will be aligned and the fill value is zero
+        (``"axis": None, "fill_value": 0}``). This assumes that if events and/or regions
+        between both data frames do not align, the respective value is assumed to be
+        zero and this will be incorporated into the estimation. If you want to require
+        alignment, set ``"fill_value": None``. This will set non-aligned values to NaN,
+        which typically results in a NaN target function, aborting the estimation.
+    assign_centroids : bool, optional
+        If ``True`` (default), assign the hazard centroids to the exposure.
+    """
+
+    hazard: Hazard
+    exposure: Exposures
+    data: pd.DataFrame
+    impact_func_creator: Callable[..., ImpactFuncSet]
+    impact_to_dataframe: Callable[[Impact], pd.DataFrame]
+    cost_func: Callable[[pd.DataFrame, pd.DataFrame], Number]
+    bounds: Optional[Mapping[str, Union[Bounds, Tuple[Number, Number]]]] = None
+    constraints: Optional[Union[ConstraintType, list[ConstraintType]]] = None
+    impact_calc_kwds: Mapping[str, Any] = field(
+        default_factory=lambda: {"assign_centroids": False}
+    )
+    align_kwds: Mapping[str, Any] = field(
+        default_factory=lambda: {"axis": None, "fill_value": 0}
+    )
+    assign_centroids: InitVar[bool] = True
+
+    def __post_init__(self, assign_centroids):
+        """Prepare input data"""
+        if assign_centroids:
+            self.exposure.assign_centroids(self.hazard)
+
+
+@dataclass
+class Output:
+    """Generic output of a calibration task
+
+    Attributes
+    ----------
+    params : Mapping (str, Number)
+        The optimal parameters
+    target : Number
+        The target function value for the optimal parameters
+    """
+
+    params: Mapping[str, Number]
+    target: Number
+
+
+@dataclass
+class OutputEvaluator:
+    """Evaluate the output of a calibration task
+
+    Parameters
+    ----------
+    input : Input
+        The input object for the optimization task.
+    output : Output
+        The output object returned by the optimization task.
+
+    Attributes
+    ----------
+    impf_set : climada.entity.ImpactFuncSet
+        The impact function set built from the optimized parameters
+    impact : climada.engine.Impact
+        An impact object calculated using the optimal :py:attr:`impf_set`
+    """
+
+    input: Input
+    output: Output
+
+    def __post_init__(self):
+        """Compute the impact for the optimal parameters"""
+        self.impf_set = self.input.impact_func_creator(**self.output.params)
+        self.impact = ImpactCalc(
```

Comment:
TODO: Calculate impact in separate method (for passing kwargs)

Reply:
I decided against that because all methods of this class require the impact data